```
Error running mint automation
+ mkdir -p 15662-f9a1525/gopath/src/github.com/minio
+ git -C 15662-f9a1525/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 15662-f9a1525/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 15662-f9a1525/gopath/src/github.com/minio/minio fetch --quiet minio pull/15662/head:pr15662
+ git -C 15662-f9a1525/gopath/src/github.com/minio/minio checkout --quiet pr15662
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/15662-f9a1525/gopath
+ make -C 15662-f9a1525/gopath/src/github.com/minio/minio --quiet
go: downloading github.com/minio/sha256-simd v1.0.0
go: downloading github.com/minio/cli v1.23.0
go: downloading github.com/json-iterator/go v1.1.12
go: downloading github.com/minio/pkg v1.3.0
go: downloading github.com/Shopify/sarama v1.35.0
go: downloading github.com/golang-jwt/jwt/v4 v4.4.1
go: downloading github.com/minio/highwayhash v1.0.2
go: downloading github.com/minio/sio v0.3.0
go: downloading github.com/minio/madmin-go v1.4.26
go: downloading github.com/klauspost/compress v1.15.9
go: downloading github.com/buger/jsonparser v1.1.1
go: downloading github.com/minio/minio-go/v7 v7.0.35
go: downloading github.com/secure-io/sio-go v0.3.1
go: downloading github.com/xdg/scram v0.0.0-20180814205039-7eeb5667e42c
go: downloading github.com/rs/dnscache v0.0.0-20211102005908-e0241e321417
go: downloading github.com/minio/kes v0.20.0
go: downloading github.com/dustin/go-humanize v1.0.0
go: downloading golang.org/x/crypto v0.0.0-20220722155217-630584e8d5aa
go: downloading github.com/fatih/color v1.13.0
go: downloading golang.org/x/sys v0.0.0-20220722155257-8c9f86f7a55f
go: downloading cloud.google.com/go/storage v1.10.0
go: downloading github.com/klauspost/readahead v1.4.0
go: downloading github.com/go-openapi/loads v0.21.1
go: downloading github.com/cheggaaa/pb v1.0.29
go: downloading github.com/google/uuid v1.3.0
go: downloading github.com/minio/console v0.20.0
go: downloading github.com/bcicen/jstream v1.0.1
go: downloading github.com/hashicorp/golang-lru v0.5.4
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/coreos/go-oidc v2.1.0+incompatible
go: downloading github.com/inconshreveable/mousetrap v1.0.0
go: downloading github.com/djherbis/atime v1.0.0
go: downloading github.com/felixge/fgprof v0.9.2
go: downloading github.com/bits-and-blooms/bloom/v3 v3.0.1
go: downloading github.com/dchest/siphash v1.2.1
go: downloading github.com/klauspost/reedsolomon v1.10.0
go: downloading github.com/miekg/dns v1.1.49
go: downloading github.com/klauspost/pgzip v1.2.5
go: downloading github.com/cespare/xxhash/v2 v2.1.2
go: downloading github.com/cosnicolaou/pbzip2 v1.0.1
go: downloading github.com/minio/dperf v0.4.2
go: downloading github.com/Azure/azure-storage-blob-go v0.10.0
go: downloading github.com/minio/selfupdate v0.5.0
go: downloading github.com/minio/xxml v0.0.3
go: downloading github.com/minio/zipindex v0.2.1
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/philhofer/fwd v1.1.2-0.20210722190033-5c56ac6d0bb9
go: downloading github.com/pierrec/lz4 v2.6.1+incompatible
go: downloading github.com/prometheus/client_golang v1.12.2
go: downloading github.com/prometheus/client_model v0.2.0
go: downloading github.com/prometheus/procfs v0.7.3
go: downloading github.com/rs/cors v1.7.0
go: downloading github.com/shirou/gopsutil/v3 v3.22.7
go: downloading github.com/valyala/bytebufferpool v1.0.0
go: downloading github.com/prometheus/common v0.34.0
go: downloading github.com/tinylib/msgp v1.1.7-0.20211026165309-e818a1881b0e
go: downloading github.com/yargevad/filepathx v1.0.0
go: downloading github.com/zeebo/xxh3 v1.0.2
go: downloading go.etcd.io/etcd/api/v3 v3.5.4
go: downloading go.etcd.io/etcd/client/v3 v3.5.4
go: downloading go.uber.org/atomic v1.9.0
go: downloading golang.org/x/oauth2 v0.0.0-20220411215720-9780585627b5
go: downloading google.golang.org/api v0.81.0
go: downloading golang.org/x/time v0.0.0-20220224211638-0e9765cccd65
go: downloading github.com/beevik/ntp v0.3.0
go: downloading go.uber.org/zap v1.21.0
go: downloading github.com/coredns/coredns v1.9.3
go: downloading github.com/minio/simdjson-go v0.4.2
go: downloading github.com/eclipse/paho.mqtt.golang v1.4.1
go: downloading github.com/elastic/go-elasticsearch/v7 v7.17.1
go: downloading github.com/go-sql-driver/mysql v1.6.0
go: downloading github.com/gomodule/redigo v1.8.8
go: downloading github.com/lib/pq v1.10.4
go: downloading github.com/nats-io/nats.go v1.16.0
go: downloading github.com/nats-io/stan.go v0.10.2
go: downloading github.com/nsqio/go-nsq v1.0.8
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/streadway/amqp v1.0.0
go: downloading github.com/ncw/directio v1.0.5
go: downloading github.com/go-ldap/ldap/v3 v3.4.3
go: downloading github.com/bits-and-blooms/bitset v1.2.0
go: downloading github.com/mattn/go-colorable v0.1.12
go: downloading github.com/mattn/go-isatty v0.0.14
go: downloading github.com/mattn/go-runewidth v0.0.13
go: downloading github.com/pquerna/cachecontrol v0.1.0
go: downloading gopkg.in/square/go-jose.v2 v2.5.1
go: downloading github.com/klauspost/cpuid/v2 v2.1.0
go: downloading github.com/minio/csvparser v1.0.0
go: downloading github.com/fraugster/parquet-go v0.10.0
go: downloading github.com/alecthomas/participle v0.2.1
go: downloading github.com/xdg/stringprep v1.0.0
go: downloading github.com/golang/protobuf v1.5.2
go: downloading golang.org/x/net v0.0.0-20220722155237-a158d28d115b
go: downloading github.com/google/pprof v0.0.0-20211214055906-6f57359322fd
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/modern-go/reflect2 v1.0.2
go: downloading go.etcd.io/etcd/client/pkg/v3 v3.5.4
go: downloading google.golang.org/grpc v1.46.2
go: downloading github.com/go-openapi/analysis v0.21.2
go: downloading github.com/go-openapi/spec v0.20.5
go: downloading github.com/go-openapi/swag v0.21.1
go: downloading github.com/golang/snappy v0.0.4
go: downloading github.com/nats-io/nuid v1.0.1
go: downloading github.com/Azure/azure-pipeline-go v0.2.2
go: downloading cloud.google.com/go v0.100.2
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.2-0.20181231171920-c182affec369
go: downloading golang.org/x/sync v0.0.0-20220513210516-0976fa681c29
go: downloading github.com/Azure/go-ntlmssp v0.0.0-20211209120228-48547f28849e
go: downloading github.com/go-asn1-ber/asn1-ber v1.5.4
go: downloading google.golang.org/genproto v0.0.0-20220519153652-3a47de7e79bd
go: downloading github.com/coreos/go-semver v0.3.0
go: downloading github.com/rivo/uniseg v0.2.0
go: downloading github.com/gorilla/websocket v1.5.0
go: downloading golang.org/x/text v0.3.7
go: downloading github.com/rs/xid v1.4.0
go: downloading gopkg.in/ini.v1 v1.66.6
go: downloading github.com/minio/md5-simd v1.1.2
go: downloading github.com/beorn7/perks v1.0.1
go: downloading google.golang.org/protobuf v1.28.0
go: downloading go.uber.org/multierr v1.8.0
go: downloading github.com/nats-io/nkeys v0.3.0
go: downloading github.com/mattn/go-ieproxy v0.0.1
go: downloading github.com/coreos/go-systemd/v22 v22.3.2
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/eapache/go-resiliency v1.3.0
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading github.com/eapache/queue v1.1.0
go: downloading github.com/jcmturner/gofork v1.0.0
go: downloading github.com/jcmturner/gokrb5/v8 v8.4.2
go: downloading github.com/hashicorp/go-multierror v1.1.1
go: downloading github.com/pierrec/lz4/v4 v4.1.15
go: downloading github.com/rcrowley/go-metrics v0.0.0-20201227073835-cf1acfcdf475
go: downloading github.com/mailru/easyjson v0.7.7
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/lestrrat-go/jwx v1.2.19
go: downloading github.com/fatih/structs v1.1.0
go: downloading github.com/montanaflynn/stats v0.6.6
go: downloading github.com/rjeczalik/notify v0.9.2
go: downloading github.com/hashicorp/errwrap v1.1.0
go: downloading github.com/tklauser/go-sysconf v0.3.10
go: downloading github.com/go-openapi/jsonpointer v0.19.5
go: downloading github.com/go-openapi/strfmt v0.21.2
go: downloading github.com/josharian/intern v1.0.0
go: downloading github.com/tklauser/numcpus v0.4.0
go: downloading github.com/oklog/ulid v1.3.1
go: downloading github.com/go-openapi/errors v0.20.2
go: downloading go.mongodb.org/mongo-driver v1.9.1
go: downloading github.com/mitchellh/mapstructure v1.4.3
go: downloading github.com/asaskevich/govalidator v0.0.0-20210307081110-f21760c49a8d
go: downloading github.com/jcmturner/dnsutils/v2 v2.0.0
go: downloading github.com/hashicorp/go-uuid v1.0.2
go: downloading github.com/jcmturner/rpc/v2 v2.0.3
go: downloading github.com/jcmturner/aescts/v2 v2.0.0
go: downloading github.com/apache/thrift v0.15.0
go: downloading github.com/go-openapi/jsonreference v0.20.0
go: downloading github.com/lestrrat-go/iter v1.0.1
go: downloading github.com/lestrrat-go/backoff/v2 v2.0.8
go: downloading github.com/lestrrat-go/option v1.0.0
go: downloading github.com/lestrrat-go/httpcc v1.0.0
go: downloading github.com/lestrrat-go/blackmagic v1.0.0
go: downloading cloud.google.com/go/iam v0.2.0
go: downloading github.com/googleapis/gax-go/v2 v2.4.0
go: downloading go.opencensus.io v0.23.0
go: downloading golang.org/x/xerrors v0.0.0-20220517211312-f3a8303e98df
go: downloading github.com/golang/groupcache v0.0.0-20210331224755-41bb18bfe9da
go: downloading github.com/go-stack/stack v1.8.1
go: downloading github.com/go-openapi/runtime v0.23.3
go: downloading github.com/go-openapi/validate v0.21.0
go: downloading github.com/jessevdk/go-flags v1.5.0
go: downloading github.com/minio/mc v0.0.0-20220818165341-8c239d16aa37
go: downloading github.com/unrolled/secure v1.10.0
go: downloading github.com/tidwall/gjson v1.14.0
go: downloading github.com/tidwall/match v1.1.1
go: downloading github.com/tidwall/pretty v1.2.0
go: downloading github.com/docker/go-units v0.4.0
go: downloading github.com/charmbracelet/bubbles v0.10.3
go: downloading github.com/google/shlex v0.0.0-20191202100458-e7afc7fbc510
go: downloading github.com/charmbracelet/lipgloss v0.5.0
go: downloading github.com/olekukonko/tablewriter v0.0.5
go: downloading github.com/rivo/tview v0.0.0-20220216162559-96063d6082f3
go: downloading github.com/prometheus/prom2json v1.3.1
go: downloading github.com/gdamore/tcell/v2 v2.4.1-0.20210905002822-f057f0a857a1
go: downloading golang.org/x/term v0.0.0-20210927222741-03fcf44c2211
go: downloading github.com/minio/filepath v1.0.0
go: downloading github.com/navidys/tvxwidgets v0.1.0
go: downloading github.com/charmbracelet/bubbletea v0.20.0
go: downloading github.com/minio/colorjson v1.0.2
go: downloading github.com/posener/complete v1.2.3
go: downloading gopkg.in/h2non/filetype.v1 v1.0.5
go: downloading github.com/pkg/xattr v0.4.5
go: downloading github.com/muesli/termenv v0.11.1-0.20220212125758-44cd13922739
go: downloading github.com/muesli/reflow v0.3.0
go: downloading github.com/lucasb-eyer/go-colorful v1.2.0
go: downloading github.com/muesli/ansi v0.0.0-20211031195517-c9f0611b6c70
go: downloading github.com/containerd/console v1.0.3
go: downloading github.com/gdamore/encoding v1.0.0
go: downloading cloud.google.com/go/compute v1.6.1
# github.com/minio/minio/cmd
cmd/generic-handlers.go:259:7: t declared but not used
cmd/generic-handlers.go:262:11: undefined: t
make: *** [build] Error 2

```

--- a/cmd/bucket-lifecycle.go
+++ b/cmd/bucket-lifecycle.go
@@ -31,6 +31,7 @@ import (
 
 	"github.com/google/uuid"
 	"github.com/minio/minio-go/v7/pkg/tags"
+	"github.com/minio/minio/internal/amztime"
 	sse "github.com/minio/minio/internal/bucket/encryption"
 	"github.com/minio/minio/internal/bucket/lifecycle"
 	"github.com/minio/minio/internal/event"
@@ -786,7 +787,7 @@ func parseRestoreObjStatus(restoreHdr string) (restoreObjStatus, error) {
 		if strings.TrimSpace(expiryTokens[0]) != "expiry-date" {
 			return restoreObjStatus{}, errRestoreHDRMalformed
 		}
-		expiry, err := time.Parse(http.TimeFormat, strings.Trim(expiryTokens[1], `"`))
+		expiry, err := amztime.ParseHeader(strings.Trim(expiryTokens[1], `"`))
 		if err != nil {
 			return restoreObjStatus{}, errRestoreHDRMalformed
 		}

--- a/cmd/disk-cache-backend.go
+++ b/cmd/disk-cache-backend.go
@@ -38,6 +38,7 @@ import (
 	"time"
 
 	"github.com/djherbis/atime"
+	"github.com/minio/minio/internal/amztime"
 	"github.com/minio/minio/internal/config/cache"
 	"github.com/minio/minio/internal/crypto"
 	"github.com/minio/minio/internal/disk"
@@ -132,17 +133,14 @@ func (m *cacheMeta) ToObjectInfo() (o ObjectInfo) {
 	} else {
 		o.StorageClass = globalMinioDefaultStorageClass
 	}
-	var (
-		t time.Time
-		e error
-	)
+
 	if exp, ok := meta["expires"]; ok {
-		if t, e = time.Parse(http.TimeFormat, exp); e == nil {
+		if t, e := amztime.ParseHeader(exp); e == nil {
 			o.Expires = t.UTC()
 		}
 	}
 	if mtime, ok := meta["last-modified"]; ok {
-		if t, e = time.Parse(http.TimeFormat, mtime); e == nil {
+		if t, e := amztime.ParseHeader(mtime); e == nil {
 			o.ModTime = t.UTC()
 		}
 	}

--- a/cmd/erasure-metadata.go
+++ b/cmd/erasure-metadata.go
@@ -26,6 +26,7 @@ import (
 	"strings"
 	"time"
 
+	"github.com/minio/minio/internal/amztime"
 	"github.com/minio/minio/internal/bucket/replication"
 	"github.com/minio/minio/internal/hash/sha256"
 	xhttp "github.com/minio/minio/internal/http"
@@ -127,17 +128,8 @@ func (fi FileInfo) ToObjectInfo(bucket, object string, versioned bool) ObjectInf
 		SuccessorModTime: fi.SuccessorModTime,
 	}
 
-	// Update expires
-	var (
-		t time.Time
-		e error
-	)
-
-	const nonStandardHTTPTimeFormat = "Mon, 2 Jan 2006 15:04:05 GMT"
 	if exp, ok := fi.Metadata["expires"]; ok {
-		if t, e = time.Parse(http.TimeFormat, exp); e == nil {
-			objInfo.Expires = t.UTC()
-		} else if t, e = time.Parse(nonStandardHTTPTimeFormat, exp); e == nil {
+		if t, err := amztime.ParseHeader(exp); err == nil {
 			objInfo.Expires = t.UTC()
 		}
 	}

--- a/cmd/fs-v1-metadata.go
+++ b/cmd/fs-v1-metadata.go
@@ -23,12 +23,11 @@ import (
 	"encoding/json"
 	"io"
 	"io/ioutil"
-	"net/http"
 	"os"
 	pathutil "path"
-	"time"
 
 	jsoniter "github.com/json-iterator/go"
+	"github.com/minio/minio/internal/amztime"
 	xhttp "github.com/minio/minio/internal/http"
 	"github.com/minio/minio/internal/lock"
 	"github.com/minio/minio/internal/logger"
@@ -175,12 +174,9 @@ func (m fsMetaV1) ToObjectInfo(bucket, object string, fi os.FileInfo) ObjectInfo
 	} else {
 		objInfo.StorageClass = globalMinioDefaultStorageClass
 	}
-	var (
-		t time.Time
-		e error
-	)
+
 	if exp, ok := m.Meta["expires"]; ok {
-		if t, e = time.Parse(http.TimeFormat, exp); e == nil {
+		if t, e := amztime.ParseHeader(exp); e == nil {
 			objInfo.Expires = t.UTC()
 		}
 	}

--- a/cmd/generic-handlers.go
+++ b/cmd/generic-handlers.go
@@ -27,10 +27,11 @@ import (
 	"sync/atomic"
 	"time"
 
+	"github.com/dustin/go-humanize"
 	"github.com/minio/minio-go/v7/pkg/set"
 	xnet "github.com/minio/pkg/net"
 
-	"github.com/dustin/go-humanize"
+	"github.com/minio/minio/internal/amztime"
 	"github.com/minio/minio/internal/config/dns"
 	"github.com/minio/minio/internal/crypto"
 	xhttp "github.com/minio/minio/internal/http"
@@ -241,17 +242,6 @@ func isAdminReq(r *http.Request) bool {
 	return strings.HasPrefix(r.URL.Path, adminPathPrefix)
 }
 
-// Supported amz date formats.
-var amzDateFormats = []string{
-	// Do not change this order, x-amz-date format is usually in
-	// iso8601Format rest are meant for relaxed handling of other
-	// odd SDKs that might be out there.
-	iso8601Format,
-	time.RFC1123,
-	time.RFC1123Z,
-	// Add new AMZ date formats here.
-}
-
 // Supported Amz date headers.
 var amzDateHeaders = []string{
 	// Do not chane this order, x-amz-date value should be
@@ -260,24 +250,17 @@ var amzDateHeaders = []string{
 	"date",
 }
 
-// parseAmzDate - parses date string into supported amz date formats.
-func parseAmzDate(amzDateStr string) (amzDate time.Time, apiErr APIErrorCode) {
-	for _, dateFormat := range amzDateFormats {
-		amzDate, err := time.Parse(dateFormat, amzDateStr)
-		if err == nil {
-			return amzDate, ErrNone
-		}
-	}
-	return time.Time{}, ErrMalformedDate
-}
-
 // parseAmzDateHeader - parses supported amz date headers, in
 // supported amz date formats.
 func parseAmzDateHeader(req *http.Request) (time.Time, APIErrorCode) {
 	for _, amzDateHeader := range amzDateHeaders {
 		amzDateStr := req.Header.Get(amzDateHeader)
 		if amzDateStr != "" {
-			return parseAmzDate(amzDateStr)
+			t, err := amztime.Parse(amzDateStr)
+			if err != nil {
+				return time.Time{}, ErrMalformedDate
+			}
+			return t, ErrNone
 		}
 	}
 	// Date header missing.

--- a/cmd/object-handlers-common.go
+++ b/cmd/object-handlers-common.go
@@ -24,6 +24,7 @@ import (
 	"strconv"
 	"time"
 
+	"github.com/minio/minio/internal/amztime"
 	"github.com/minio/minio/internal/event"
 	"github.com/minio/minio/internal/hash"
 	xhttp "github.com/minio/minio/internal/http"
@@ -78,7 +79,7 @@ func checkCopyObjectPreconditions(ctx context.Context, w http.ResponseWriter, r 
 	// since the specified time otherwise return 412 (precondition failed).
 	ifModifiedSinceHeader := r.Header.Get(xhttp.AmzCopySourceIfModifiedSince)
 	if ifModifiedSinceHeader != "" {
-		if givenTime, err := time.Parse(http.TimeFormat, ifModifiedSinceHeader); err == nil {
+		if givenTime, err := amztime.ParseHeader(ifModifiedSinceHeader); err == nil {
 			if !ifModifiedSince(objInfo.ModTime, givenTime) {
 				// If the object is not modified since the specified time.
 				writeHeaders()
@@ -92,7 +93,7 @@ func checkCopyObjectPreconditions(ctx context.Context, w http.ResponseWriter, r 
 	// modified since the specified time, otherwise return a 412 (precondition failed).
 	ifUnmodifiedSinceHeader := r.Header.Get(xhttp.AmzCopySourceIfUnmodifiedSince)
 	if ifUnmodifiedSinceHeader != "" {
-		if givenTime, err := time.Parse(http.TimeFormat, ifUnmodifiedSinceHeader); err == nil {
+		if givenTime, err := amztime.ParseHeader(ifUnmodifiedSinceHeader); err == nil {
 			if ifModifiedSince(objInfo.ModTime, givenTime) {
 				// If the object is modified since the specified time.
 				writeHeaders()
@@ -172,7 +173,7 @@ func checkPreconditions(ctx context.Context, w http.ResponseWriter, r *http.Requ
 	// otherwise return a 304 (not modified).
 	ifModifiedSinceHeader := r.Header.Get(xhttp.IfModifiedSince)
 	if ifModifiedSinceHeader != "" {
-		if givenTime, err := time.Parse(http.TimeFormat, ifModifiedSinceHeader); err == nil {
+		if givenTime, err := amztime.ParseHeader(ifModifiedSinceHeader); err == nil {
 			if !ifModifiedSince(objInfo.ModTime, givenTime) {
 				// If the object is not modified since the specified time.
 				writeHeaders()
@@ -186,7 +187,7 @@ func checkPreconditions(ctx context.Context, w http.ResponseWriter, r *http.Requ
 	// time, otherwise return a 412 (precondition failed).
 	ifUnmodifiedSinceHeader := r.Header.Get(xhttp.IfUnmodifiedSince)
 	if ifUnmodifiedSinceHeader != "" {
-		if givenTime, err := time.Parse(http.TimeFormat, ifUnmodifiedSinceHeader); err == nil {
+		if givenTime, err := amztime.ParseHeader(ifUnmodifiedSinceHeader); err == nil {
 			if ifModifiedSince(objInfo.ModTime, givenTime) {
 				// If the object is modified since the specified time.
 				writeHeaders()

--- a/internal/amztime/parse.go
+++ b/internal/amztime/parse.go
@@ -1,0 +1,72 @@
+// Copyright (c) 2015-2022 MinIO, Inc.
+//
+// This file is part of MinIO Object Storage stack
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// Package amztime implements AWS specific time parsing and deviations
+package amztime
+
+import (
+	"errors"
+	"net/http"
+	"time"
+)
+
+// Supported amz date formats.
+var amzDateFormats = []string{
+	// Do not change this order, x-amz-date format is usually in
+	// iso8601Format rest are meant for relaxed handling of other
+	// odd SDKs that might be out there.
+	"20060102T150405Z",
+	time.RFC1123,
+	time.RFC1123Z,
+	// Add new AMZ date formats here.
+}
+
+// ErrMalformedDate always returned for dates that cannot be parsed.
+var ErrMalformedDate = errors.New("malformed date")
+
+// Parse parses date string via supported amz date formats.
+func Parse(amzDateStr string) (time.Time, error) {
+	for _, dateFormat := range amzDateFormats {
+		amzDate, err := time.Parse(dateFormat, amzDateStr)
+		if err == nil {
+			return amzDate, nil
+		}
+	}
+	return time.Time{}, ErrMalformedDate
+}
+
+var httpTimeFormats = []string{
+	// Do not chagne this order, http time format dates
+	// are usually in http.TimeFormat however there are
+	// situations where for example aws-sdk-java doesn't
+	// send the correct format.
+	http.TimeFormat,
+	"Mon, 2 Jan 2006 15:04:05 GMT",
+}
+
+// ParseHeader parses http.TimeFormat with an acceptable
+// extension for http.TimeFormat - return time might be zero
+// if the timeStr is invalid.
+func ParseHeader(timeStr string) (time.Time, error) {
+	for _, dateFormat := range httpTimeFormats {
+		t, err := time.Parse(dateFormat, timeStr)
+		if err == nil {
+			return t, nil
+		}
+	}
+	return time.Time{}, ErrMalformedDate
+}

--- a/internal/amztime/parse_test.go
+++ b/internal/amztime/parse_test.go
@@ -1,0 +1,58 @@
+// Copyright (c) 2015-2022 MinIO, Inc.
+//
+// This file is part of MinIO Object Storage stack
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// Package amztime implements AWS specific time parsing and deviations
+package amztime
+
+import (
+	"errors"
+	"testing"
+	"time"
+)
+
+func TestParse(t *testing.T) {
+	type testCase struct {
+		expectedErr  error
+		expectedTime time.Time
+		timeStr      string
+	}
+	testCases := []testCase{
+		{
+			ErrMalformedDate,
+			time.Time{},
+			"Tue Sep  6 07:10:23 PM PDT 2022",
+		},
+		{
+			nil,
+			time.Date(2009, time.November, 10, 23, 0, 0, 0, time.UTC),
+			"Tue, 10 Nov 2009 23:00:00 UTC",
+		},
+	}
+
+	for _, testCase := range testCases {
+		testCase := testCase
+		t.Run(testCase.timeStr, func(t *testing.T) {
+			gott, goterr := Parse(testCase.timeStr)
+			if !errors.Is(goterr, testCase.expectedErr) {
+				t.Errorf("expected %v, got %v", testCase.expectedErr, goterr)
+			}
+			if !gott.Equal(testCase.expectedTime) {
+				t.Errorf("expected %v, got %v", testCase.expectedTime, gott)
+			}
+		})
+	}
+}

